File: [47] Permutations II/47.permutations-ii.cpp
/*
 * @lc app=leetcode id=47 lang=cpp
 *
 * [47] Permutations II
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        // this method swap nums position
        vector<vector<int>> res;
        DFS(nums, res, 0);
        return res;
    }

    // permute num[start..end]
    // invariant: num[0..start-1] have been fixed/permuted
    void DFS(vector<int>& nums, vector<vector<int>>& res, int start) {
        if (start == nums.size()) {
            res.emplace_back(nums);
            return;
        } else {
            for (int i = start; i < nums.size(); i++) {
                int j = i-1;
                //judge whether num[i] and num[start], ..., nums[i-1] are equal
                while(j >= start && nums[j] != nums[i]) {
                    j--;
                }
                if (j == start-1) {
                    swap(nums[start], nums[i]);
                    DFS(nums, res, start+1);
                    // reset
                    swap(nums[start], nums[i]);
                }
            }
        }
    }
};
// @lc code=end

File: [47] Permutations II/47.permutations-ii.java
import java.util.ArrayList;
import java.util.List;

/*
 * @lc app=leetcode id=47 lang=java
 *
 * [47] Permutations II
 */

// @lc code=start
class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(nums);
        DFS(nums, res, new ArrayList<Integer>(), new boolean[nums.length]);
        return res;
    }

    public void DFS(int[] nums, List<List<Integer>> res, List<Integer> temp_list, boolean[] used) {
        if (temp_list.size() == nums.length) {
            res.add(new ArrayList<>(temp_list));
            return;
        } else {
            for (int i = 0; i < nums.length; i++) {
                if (used[i]) continue;
                // this condition is for avoiding repeated [1,1,2] fot the test case
                // In fact, both !used[i-1] and used[i-1] work
                // !used[i-1] makes sure when duplicates are selected, the order is ascending (index from small to large). 
                // However, used[i-1] means the descending order.
                if (i != 0 && nums[i] == nums[i-1] && !used[i-1]) continue;
                temp_list.add(nums[i]);
                used[i] = true;
                DFS(nums, res, temp_list, used);
                used[i] = false;
                temp_list.remove(temp_list.size()-1);
            }
        }
    }
}
// @lc code=end

